su/su.1
.\"
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.\" Sccsid @(#)su.1	1.10 (gritter) 4/17/03
.TH SU 1 "4/17/03" "Heirloom Toolchest" "User Commands"
.SH NAME
su \- become super-user or another user
.SH SYNOPSIS
\fBsu\fR [\fB\-\fR] [\fIusername\fR [\fIarg ...\fR]]
.SH DESCRIPTION
The
.B su
command allows one user to get the credentials of another
without logging on and off.
The default
.I username
is
.B root
(that is, the
.IR super-user ).
.PP
Unless the caller is the super-user,
the user's password must be entered
(the exact authentication mechanism may depend on
.SM PAM
settings for
.BR su ,
see
.IR pam (8)).
If access is granted, the
.B su
command will execute the user's shell from the system's password file
with the appropriate user id, group id and supplementary group ids.
Any additional arguments will be passed to this shell.
.PP
If the first argument is
.BR \- ,
a login shell is executed in the user's home directory
using default environment variables;
only the settings of the
.SM DISPLAY
and
.SM TERM
variables are retained.
In any case, the
.SM PATH
environment variable is adjusted.
.PP
The
.B su
command reads the configuration file
.B /etc/default/su
on startup. Lines containing the following strings are interpreted:
.TP
.BI CONSOLE= filename
All attempts are logged to the specified file, usually
.BR /dev/console .
If no such line is present, console logging is not done.
.TP
.BI PATH= path
The default path for non-super-user accounts. If unset,
.I /usr/local/bin:/bin:/usr/bin:
is assumed.
.TP
.BI SLEEPTIME= seconds
If authentication fails,
.B su
will wait the specified number of seconds
before printing an error message and exiting.
Default is 4, minimum is 0, maximum is 5 seconds.
.TP
.BI SULOG= filename
All attempts are logged to the specified file, if present.
.TP
.BI SUPATH= path
.ad l
The default path for super-user accounts. If unset,
.I /usr/local/sbin:/usr/local/bin:/sbin:/usr/sbin:/bin:/usr/bin
is assumed.
.ad b
.TP
.BI SYSLOG= value
Log attempts to the
.IR syslog (3)
.I LOG_AUTH
facility. If
.I value is
.IR YES ,
all attempts are logged;
if
.IR FAIL ,
only the ones that failed.
.I LOG_CRIT
messages are generated for failed
.B su
attempts,
.I LOG_NOTICE
messages for successful attempts to become super-user, and
.I LOG_INFO
messages otherwise.
.SH FILES
.TP
.B /etc/default/su
Configuration file.
.TP
.B /etc/passwd
User database.
.SH "SEE ALSO"
env(1),
login(1),
sh(1),
syslog(3),
passwd(5),
pam(8)


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

su/su.c
/*
 * su - become-super user or another user
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, May 2001.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)su.sl	1.25 (gritter) 2/21/06";

#include	"config.h"
#include	<sys/types.h>
#include	<sys/wait.h>
#include	<sys/stat.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<strings.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<libgen.h>
#include	<time.h>
#include	<signal.h>
#include	<pwd.h>
#include	<grp.h>
#include	<termios.h>
#include	<limits.h>
#include	<syslog.h>
#ifdef	__APPLE__
#include	<pam/pam_appl.h>
#elif	PAM
#include	<security/pam_appl.h>
#else	/* !PAM */
#ifdef SHADOW_PWD
#include	<shadow.h>
#endif	/* SHADOW_PWD */
#define	PAM_MAX_RESP_SIZE	512
#endif	/* !PAM */

enum logtype {
	LT_NONE,
	LT_FAIL,
	LT_ALL
};

extern char	**environ;

static int	dash;			/* create login environment */
static int	dofork;			/* fork() before executing shell */
static enum logtype	dosyslog = LT_NONE;	/* log to syslog */
static int	sleeptime = 4;		/* sleep time if failed */
static char	*user;			/* desired login name */
static char	*olduser;		/* old login name */
static char	*progname;		/* argv[0] to main() */
static char	*path;			/* new $PATH */
static char	*dir;			/* new home directory */
static char	*shell;			/* new shell */
static char	*sulog;			/* logging file */
static char	*console;		/* console to log to (optional) */
static struct passwd	*pwd;			/* passwd entry for new user */
static void	(*oldint)(int);		/* old SIGINT handler */
static void	(*oldquit)(int);	/* old SIGQUIT handler */

static const char	defaultfile[] = SUDFL;
static const char	PATH[] = "PATH=/usr/local/bin:/bin:/usr/bin:";
static const char	SUPATH[] =
	"PATH=/usr/local/sbin:/usr/local/bin:/sbin:/usr/sbin:/bin:/usr/bin";

/*
 * Memory allocation with check.
 */
static void *
smalloc(size_t nbytes)
{
	void *p;

	if ((p = malloc(nbytes)) == NULL) {
		write(2, "Out of memory\n", 14);
		exit(077);
	}
	return p;
}

/*
 * Adjust environment.
 */
static void
doenv(void)
{

	if (dash) {
		char *cp;
		char *term, *oldterm;
		char *display, *olddisplay;

		if ((oldterm = getenv("TERM")) != NULL) {
			term = smalloc(strlen(oldterm) + 6);
			sprintf(term, "TERM=%s", oldterm);
		} else
			term = NULL;
		if ((olddisplay = getenv("DISPLAY")) != NULL) {
			display = smalloc(strlen(olddisplay) + 9);
			sprintf(display, "DISPLAY=%s", olddisplay);
		} else
			display = NULL;
		environ = NULL;
		cp = smalloc(strlen(dir) + 6);
		sprintf(cp, "HOME=%s", dir);
		putenv(cp);
		cp = smalloc(strlen(user) + 9);
		sprintf(cp, "LOGNAME=%s", user);
		putenv(cp);
		cp = smalloc(strlen(shell) + 7);
		sprintf(cp, "SHELL=%s", shell);
		putenv(cp);
		if (term)
			putenv(term);
		if (display)
			putenv(display);
	}
	putenv(path);
	if (pwd->pw_uid == 0)
		putenv("PS1=# ");
}

/*
 * Things to do in child process.
 */
static void
child(char **args)
{
	if (initgroups(user, pwd->pw_gid) < 0 || setgid(pwd->pw_gid) < 0
			|| setuid(pwd->pw_uid) < 0) {
		fprintf(stderr, "%s: Invalid ID\n", progname);
		dofork ? _exit(1) : exit(1);
	}
	if (dash) {
		args[0] = smalloc(strlen(basename(shell)) + 2);
		args[0][0] = '-';
		strcpy(&args[0][1], basename(shell));
		if (dir == NULL || chdir(dir) < 0) {
			fprintf(stderr, "%s: No directory! Using home=/\n",
					progname);
			dir = "/";
			chdir(dir);
		}
	} else
		args[0] = basename(shell);
	doenv();
	signal(SIGINT, oldint);
	signal(SIGQUIT, oldquit);
	execv(shell, args);
	fprintf(stderr, "%s: No shell\n", progname);
	dofork ? _exit(3) : exit(3);
}

/*
 * Spawn the child process and wait for it.
 */
static int
dospawn(char **args)
{
	int status;
	pid_t pid;

	switch (pid = fork()) {
	case 0:
		child(args);
		/*NOTREACHED*/
	case -1:
		fprintf(stderr, "%s: fork() failed, try again later\n",
				progname);
		return 1;
	}
	while (wait(&status) != pid);
	return status;
}

/*
 * Write log entries.
 */
static void
dolog(int sign)
{
	FILE *fp;
	char *tty;
	const char format[] = "SU %02u/%02u %02u:%02u %c %s %s-%s\n";
	struct tm *tm;
	time_t t;

	time(&t);
	tm = localtime(&t);
	if ((tty = ttyname(0)) == NULL)
		tty = "/dev/???";
	if (dosyslog == LT_ALL || (dosyslog == LT_FAIL && sign == '-')) {
		openlog("su", LOG_PID, LOG_AUTH);
		if (sign == '-')
			syslog(LOG_CRIT | LOG_AUTH,
					"'su %s' failed for %s on %s",
				user, olduser, tty);
		else if (dosyslog != LT_FAIL)
			syslog((pwd->pw_uid ? LOG_INFO : LOG_NOTICE) | LOG_AUTH,
					"'su %s' succeeded for %s on %s",
				user, olduser, tty);
		closelog();
	}
	if (strncmp(tty, "/dev/", 5) == 0)
		tty += 5;
	if (sulog && (fp = fopen(sulog, "a+")) != NULL) {
		fprintf(fp, format, tm->tm_mon + 1, tm->tm_mday,
				tm->tm_hour, tm->tm_min,
				sign, tty,
				olduser, user);
		fclose(fp);
	}
	if (console && (fp = fopen(console, "a+")) != NULL) {
		fprintf(fp, format, tm->tm_mon + 1, tm->tm_mday,
				tm->tm_hour, tm->tm_min,
				sign, tty,
				olduser, user);
		fclose(fp);
	}
}

/*
 * Parse defaults file.
 */
static void
defaults(void)
{
	FILE *fp;

	if ((fp = fopen(defaultfile, "r")) != NULL) {
		char buf[LINE_MAX];
		char *cp;

		while (fgets(buf, sizeof buf, fp) != NULL) {
			if (buf[0] == '\0' || buf[0] == '\n' || buf[0] == '#')
				continue;
			if ((cp = strchr(buf, '\n')) != NULL)
				*cp = '\0';
			if (strncmp(buf, pwd->pw_uid ? "PATH=" : "SUPATH=",
					pwd->pw_uid ? 5 : 7) == 0) {
				path = smalloc(strlen(buf) + 1);
				strcpy(path, pwd->pw_uid ? buf : &buf[2]);
			} else if (strncmp(buf, "SULOG=", 6) == 0) {
				sulog = smalloc(strlen(buf) - 5);
				strcpy(sulog, &buf[6]);
			} else if (strncmp(buf, "CONSOLE=", 8) == 0) {
				console = smalloc(strlen(buf) - 7);
				strcpy(console, &buf[8]);
			} else if (strncmp(buf, "SLEEPTIME=", 10) == 0) {
				sleeptime = atoi(&buf[10]);
				if (sleeptime < 0 || sleeptime > 5)
					sleeptime = 4;
			} else if (strncmp(buf, "SYSLOG=", 7) == 0) {
				if (strcasecmp(&buf[7], "yes") == 0)
					dosyslog = LT_ALL;
				else if (strcasecmp(&buf[7], "fail") == 0)
					dosyslog = LT_FAIL;
			} else if (strncmp(buf, "DOFORK=", 7) == 0) {
				if (strcasecmp(&buf[7], "yes") == 0)
					dofork = 1;
			}
		}
		fclose(fp);
	}
	if (path == NULL)
		path = (char *)(pwd->pw_uid ? PATH : SUPATH);
}

/*
 * Ask for passwords.
 */
static char *
ask(const char *msg, int echo)
{
	struct termios tio;
	char *rsp;
	tcflag_t lflag;
	int fd, i = 0;
	char c;
	size_t sz;

	if ((fd = open("/dev/tty", O_RDWR)) < 0)
		fd = 0;
	if (tcgetattr(fd, &tio) < 0) {
		if (fd)
			close(fd);
		return NULL;
	}
	lflag = tio.c_lflag;
	if (echo == 0) {
		tio.c_lflag &= ~(ECHO | ECHOE | ECHOK | ECHONL);
		if (tcsetattr(fd, TCSAFLUSH, &tio) < 0) {
			if (fd)
				close(fd);
			return NULL;
		}
	}
	sz = strlen(msg);
	if (sz > 2 && msg[sz-1] == ' ' && msg[sz-2] == ':')
		sz--;
	fwrite(msg, sizeof *msg, sz, stderr);
	rsp = smalloc(PAM_MAX_RESP_SIZE + 1);
	while (read(fd, &c, 1) == 1 && c != '\n')
		if (i < PAM_MAX_RESP_SIZE)
			rsp[i++] = c;
	rsp[i] = '\0';
	fputc('\n', stderr);
	if (echo == 0) {
		tio.c_lflag = lflag;
		tcsetattr(fd, TCSADRAIN, &tio);
	}
	if (fd)
		close(fd);
	return rsp;
}

#ifdef	PAM
/*
 * PAM conversation.
 */
static int
doconv(int num_msg, const struct pam_message **msg,
		struct pam_response **rsp, void *appdata) {
	char *cp;
	int i;

	if (num_msg <= 0)
		return PAM_CONV_ERR;
	for (i = 0; i < num_msg; i++) {
		switch (msg[i]->msg_style) {
		case PAM_PROMPT_ECHO_ON:
		case PAM_PROMPT_ECHO_OFF:
			if ((cp = ask(msg[i]->msg, msg[i]->msg_style
					== PAM_PROMPT_ECHO_ON)) == NULL)
				return PAM_CONV_ERR;
			rsp[i] = smalloc(sizeof *rsp[i]);
			rsp[i]->resp = cp;
			rsp[i]->resp_retcode = 0;
			break;
		case PAM_ERROR_MSG:
			fprintf(stderr, "%s\n", msg[i]->msg);
			break;
		case PAM_TEXT_INFO:
			printf("%s\n", msg[i]->msg);
			break;
		}
	}
	return PAM_SUCCESS;
}

static struct pam_conv conv = {
	doconv,
	NULL
};
#else	/* !PAM */
/*
 * Check if the passed string is possibly a valid password, either a
 * traditional or a MD5 one.
 */
#ifdef	SHADOW_PWD
static int
cantbevalid(const char *ncrypt)
{
	if (ncrypt == NULL || strlen(ncrypt) < 13)
		return 1;
	while (*ncrypt)
		if (strchr(
	"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789./$",
				*ncrypt) == NULL)
			return 1;
	return 0;
}
#endif	/* SHADOW_PWD */

/*
 * Traditional password file authentication.
 */
static int
authenticate(const char *name, const char *ncrypt)
{
	char	*password;
	int	val;

#ifdef	SHADOW_PWD
	if (cantbevalid(ncrypt)) {
		struct spwd	*sp;
		time_t	now;

		if ((sp = getspnam(name)) == NULL)
			return 0;
		time(&now);
		if (sp->sp_expire > 0 && now >= sp->sp_expire)
			return 0;
		if (sp->sp_lstchg > 0 && sp->sp_max >= 0 && sp->sp_inact >= 0 &&
			    now >= sp->sp_lstchg + sp->sp_max + sp->sp_inact)
			return 0;
		ncrypt = sp->sp_pwdp;
	}
#endif	/* SHADOW_PWD */
	if (*ncrypt == '\0')
		return 1;
	if ((password = ask("Password:", 0)) == NULL)
		return 0;
	val = strcmp(ncrypt, crypt(password, ncrypt)) == 0;
	while (*password)
		*password++ = '\0';
	return val;
}
#endif	/* !PAM */

int
main(int argc, char **argv)
{
#ifdef	PAM
	pam_handle_t *pamh = NULL;
	int ret;
#endif	/* PAM */
	int status = 0;
	uid_t	myuid;

	progname = basename(argv[0]);
	oldint = signal(SIGINT, SIG_IGN);
	oldquit = signal(SIGQUIT, SIG_IGN);
	if (argc > 1 && argv[1][0] == '-') {
		dash++;
		argc--, argv++;
	}
	if (argc > 1) {
		user = argv[1];
		argc--, argv++;
	} else
		user = "root";
	myuid = getuid();
	if ((pwd = getpwuid(myuid)) == NULL) {
		fprintf(stderr, "%s: you do not exist\n", progname);
		exit(1);
	}
	olduser = smalloc(strlen(pwd->pw_name) + 1);
	strcpy(olduser, pwd->pw_name);
	if ((pwd = getpwnam(user)) == NULL) {
		fprintf(stderr, "%s: Unknown id: %s\n", progname, user);
		exit(1);
	}
	if (pwd->pw_dir != NULL && *pwd->pw_dir != '\0') {
		dir = smalloc(strlen(pwd->pw_dir) + 1);
		strcpy(dir, pwd->pw_dir);
	}
	if (pwd->pw_shell != NULL && *pwd->pw_shell != '\0') {
		shell = smalloc(strlen(pwd->pw_shell) + 1);
		strcpy(shell, pwd->pw_shell);
	} else
		shell = "/bin/sh";
	defaults();
#ifdef	PAM
	ret = pam_start("su", user, &conv, &pamh);
	if (ret == PAM_SUCCESS)
		ret = pam_authenticate(pamh, 0);
	if (ret == PAM_SUCCESS)
		ret = pam_acct_mgmt(pamh, 0);
	if (ret == PAM_SUCCESS) {
#else	/* PAM */
	if (myuid == 0 || authenticate(user, pwd->pw_passwd) != 0) {
#endif	/* !PAM */
		dolog('+');
		if (dofork) {
			status = dospawn(argv);
		} else {
#ifdef	PAM
			pam_end(pamh, ret);
#endif	/* PAM */
			child(argv);
			/*NOTREACHED*/
		}
	} else {
		dolog('-');
		if (sleeptime)
			sleep(sleeptime);
		fprintf(stderr, "%s: Sorry\n", progname);
		status = 1;
	}
#ifdef	PAM
	pam_end(pamh, ret);
#endif	/* PAM */
	return status;
}
